tools/pd-ctl/pdctl/command/global.go

```diff
@@ -55,69 +55,122 @@ func InitHTTPSClient(CAPath, CertPath, KeyPath string) error {
 	return nil
 }
 
-func getRequest(cmd *cobra.Command, prefix string, method string, bodyType string, body io.Reader) (*http.Request, error) {
-	if method == "" {
-		method = http.MethodGet
+type bodyOption struct {
+	contentType string
+	body        io.Reader
+}
+
+// BodyOption sets the type and content of the body
+type BodyOption func(*bodyOption)
+
+// WithBody returns a BodyOption
+func WithBody(contentType string, body io.Reader) BodyOption {
+	return func(bo *bodyOption) {
+		bo.contentType = contentType
+		bo.body = body
 	}
-	url := getAddressFromCmd(cmd, prefix)
-	req, err := http.NewRequest(method, url, body)
-	if err != nil {
-		return nil, err
+}
+func doRequest(cmd *cobra.Command, prefix string, method string,
+	opts ...BodyOption) (string, error) {
+	b := &bodyOption{}
+	for _, o := range opts {
+		o(b)
 	}
-	req.Header.Set("Content-Type", bodyType)
-	return req, err
+	each := RoundRobinURLs(cmd)
+	var resp string
+	var err error
+	each(func(endpoint string) error {
```

Comment:
I prefer something like `func tryURLs(cmd, func(endpoint string) error) error`

Reply:
As you wish :)

Reply:
I added the `tryURLs` function, it is a wrapper of `roundRobinURLs`. I did not replace `roundRobinURLs` because it is more flexible to return a closure.

The closure returned here could be used separately with its creator and can be used multiple times. The fact is we parse the command option every time to send a request which I think is unnecessary, given the fact, the design to return a closure does not show its flexibility. However, the design obeys the principle to supply the lower level APIs and then combines them to a higher level API, it will show its flexibility when we want to decouple the logic to send a request with parsing command options.

Looking for your suggestions.

Reply:
I understand your point, but I think we should either take good advantage of the `roundRobinURLs` or simply don't add it. Or we are introducing unnecessary complexity.
Maybe you can save it to a global variable so it can somehow be used in the interactive (`pd-ctl -i`) mode.

Reply:
I don't like the global var idea, because the var cannot be initialized where we declared it, we can only parse the command option when running `pd-ctl`, the separation of declaration and initialization has a bad smell to me.

It is necessary to only initialize once in the interactive mode, but I think it is hard to achieve this goal in a graceful design now. Maybe we should refactor pd-ctl later?

> or simply don't add it

I think the `roundRobinURLs` function is necessary even you change its name to `tryURLs`, the implementation cannot be avoided.  Now the `tryURLs` is a utility function to use the closure immediately after  `roundRobinURLs`(its creator), I think it is a good abstract for users.

Reply:
em, to avoid the closure, how about simply use a for loop in `tryURLs`?
```
endpoints := ...
for _, endpoint := range endpoints {
   if err := f(endpoint); err == nil {
      return nil
   }
}
return lasterr
```

Reply:
Why try to avoid it?

In my opinion: roundRobinURLs is a low-level API and tryURLs is a higher level API. User can use a higher level API to gain convenience and also can use a lower level API to gain flexibility. This does not increase the complexity because the higher level API is just a wrapper of the lower level API. If we omit the closure we will lose flexibility.

I got your point suddenly when I was writing this text. You try to avoid it because we DO NOT need it NOW, maybe we need it later but no one can tell, it depends.

OK, I will try to omit the closure.